### PR TITLE
use immutability for the NodeSet

### PR DESCRIPTION
**Pull request by paulbellamy**
Part of #502, #854

Gives a small performance boost (~30%) on the HostRenderer report.json.

### CODE REVIEW

**Comment by tomwilkie** [comment]
I'm surprised this results in such a big performance increase, but I guess its mainly due to the cheap copy.  LGTM.

File: report/node_set.go [hunk 1]
```diff
@@ -1,94 +1,197 @@
...
-		})
-		if i < len(n) && n[i].Topology == node.Topology && n[i].ID == node.ID {
-			// The list already has the element.
-			continue
-		}
-		// It a new element, insert it in order.
-		n = append(n, Node{})
-		copy(n[i+1:], n[i:])
-		n[i] = node
+		result = result.Set(fmt.Sprintf("%s %s", node.Topology, node.ID), node)
```
Comment (tomwilkie) [inline report/node_set.go]: You could consider using something other that space for the key...

File: report/node_set.go [hunk 2]
```diff
@@ -1,94 +1,197 @@
...
+	"github.com/mndrix/ps"
 )
 
-// NodeSet is a sorted set of nodes keyed on (Topology, ID). Clients must use
+// NodeSet is a set of nodes keyed on (Topology, ID). Clients must use
 // the Add method to add nodes
-type NodeSet []Node
+type NodeSet struct {
+	psMap ps.Map
+}
```
Comment (tomwilkie) [inline report/node_set.go]: Why is NodeSet a set of nodes and not a set of ids?  I thought this was just used for children?

**Comment by tomwilkie** [comment]
Also, shouldn't NodeSet be in render/?

File: report/node_set.go (hunk 2 above)
Comment (paulbellamy) [inline report/node_set.go]: It is used for children. But unlike parents, we summarise children, so we have to use the aggregated nodes. This way we store the nodes as we are "rolling up" the aggregations.

File: report/node_set.go (hunk 1 above)
Comment (paulbellamy) [inline report/node_set.go]: Sure. It doesn't really matter, aside from sorting.

**Comment by paulbellamy** [comment]
> Also, shouldn't NodeSet be in render/?

It is `Set<report.Node>`, so I felt it should live in `report`.  If it were a `Set<render.RenderableNode>`, then I'd say it should live in `render`.

File: report/node_set.go (hunk 2 above)
Comment (tomwilkie) [inline report/node_set.go]: Ah yes good point.

**Comment by tomwilkie** [comment]
> It is Set<report.Node>, so I felt it should live in report. If it were a Set<render.RenderableNode>, then I'd say it should live in render.

Ah yes we had that discussion before.  Sorry I forgot.

**Pull request merged**